src/Microsoft.AspNet.Mvc.ModelBinding/ValueProviders/BindingSourceValueProvider.cs

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNet.Mvc.ModelBinding
+{
+
+    /// <summary>
+    /// A value provider which provides data from a specific <see cref="BindingSource"/>.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// A <see cref="BindingSourceValueProvider"/> is an <see cref="IValueProvider"/> base-implementation which
+    /// can provide data for all parameters and model properties which specify the corresponding
+    /// <see cref="ModelBinding.BindingSource"/>.
+    /// </para>
+    /// <para>
+    /// <see cref="BindingSourceValueProvider"/> implements <see cref="IBindingSourceValueProvider"/> and will
+    /// include or exclude itself from the set of value providers based on the model's associated
+    /// <see cref="ModelBinding.BindingSource"/>. Value providers are by-default included; if a model does not
+    /// specify a <see cref="ModelBinding.BindingSource"/> then all value providers are valid.
+    /// </para>
+    /// </remarks>
+    public abstract class BindingSourceValueProvider : IBindingSourceValueProvider
```

Comment:
cleaned up and renamed from `MetadataAwareValueProvider`. Adios generics everywhere